### ISSUE TITLE
Make clippy less noisy when asserting in const fn context

### ISSUE THREAD

**Issue by StackOverflowExcept1on** [rust-lang/rust-clippy#11963]
### Description

This code gets 2 warnings:
1. constant is not used even if it is actually used in the const `assert!()`
2. clippy found that the assertion is always true and can be optimized, although this code can be used as a replacement for `static_assertions` (which has not been updated for 4 years)

```rust
const X: usize = 5;

#[allow(dead_code, clippy::assertions_on_constants)]
const _: () = assert!(X % 5 == 0);
// ^ equivalent of static_assertions
```

**Comment by Jarcho** [comment]
`dead_code` is a rustc lint. The second part is fixed by #11966.

**Comment by StackOverflowExcept1on** [comment]
Ok, I'll reopen issue in rust-lang/rust repo when clippy will be synced

**Comment by y21** [comment]
That `dead_code` one looks like rust-lang/rust#118424
